ligo/main.mligo

```diff
@@ -207,21 +225,118 @@ let set_position (s : storage) (i_l : tick_index) (i_u : tick_index) (i_l_l : ti
       }
     )
 
-let get_time_weighted_sum (s : storage) (c : views contract) : result =
-    ([Tezos.transaction (IC_sum s.time_weighted_ic_sum) 0mutez c], s)
+// Allocate slots in a buffer in [start, to_stop) indices range.
+// These slots will be filled with dummy values.
+let rec allocate_buffer_slots (buffer, start, to_stop : timed_cumulatives_buffer * nat * nat) : timed_cumulatives_buffer =
+    if start >= to_stop
+    then buffer
+    else
+        let new_buffer = {buffer with map = Big_map.add start dummy_timed_cumulatives buffer.map}
+        in allocate_buffer_slots(new_buffer, start + 1n, to_stop)
+
+// Increase the number of stored accumulators.
+let increase_observation_count (s, new_val : storage * nat) : result =
+    let buffer = s.cumulatives_buffer in
+    if buffer.reserved_length >= new_val
+    then (([] : operation list), s)
+    else
+        let new_buffer = allocate_buffer_slots(buffer, buffer.last + 1n, buffer.first + buffer.reserved_length)
+        in (([] : operation list), {s with cumulatives_buffer = new_buffer})
+
+// Recursive helper for `get_cumulatives`
+let rec find_cumulatives_around (buffer, t, l_i, r_i, l_v, r_v : timed_cumulatives_buffer * timestamp * nat * nat * timed_cumulatives * timed_cumulatives) : (timed_cumulatives * timed_cumulatives * nat) =
+    // Binary search, invariant: buffer[l].time <= t && t < buffer[r].time
+    if l_i + 1n < r_i
+    then
+        let m_i = (l_i + r_i) / 2n in
+        let m_v = get_registered_cumulatives_unsafe buffer m_i in
+        if m_v.time > t
+        then find_cumulatives_around (buffer, t, l_i, m_i, l_v, m_v)
+        else find_cumulatives_around (buffer, t, m_i, r_i, m_v, r_v)
+    else
+        (l_v, r_v, assert_nat (t - l_v.time, internal_observe_bin_search_failed))
```

Comment:
Can this error ever happen? :thinking:

Reply:
I suppose no, due to the invariant above :thinking: 

(Btw, I noticed a mistake in how do I handle invariants, pushed a fixed).

Reply:
In that case I suppose we could use `abs` here instead of `assert_nat`

Reply:
Hm, we already have a lot of `assert_nat(*, internal_err)`, and this one is far not the trivial one (relatively likely to occur due to a bug).

I propose handling all such `assert_nat`s altogether if we decide to use `abs` in such places.

Reply:
I see, you have a good point.

If there is a decent chance that this will be thrown due to a contract bug it's better to leave it as an `assert_nat` :ok_hand: